namer/namer.cc

```diff
@@ -1380,50 +1380,62 @@ class SymbolDefiner {
-        ENFORCE(oldDefHash.nameHash.isDefined(), "Can't delete rename if old hash is not defined");
+    bool matchesFullNameHash(core::Context ctx, core::NameRef memberNameToHash, core::FullNameHash oldNameHash) {
+        if (memberNameToHash.kind() == core::NameKind::UNIQUE) {
+            auto &uniqueData = memberNameToHash.dataUnique(ctx);
+            if (uniqueData->uniqueNameKind == core::UniqueNameKind::MangleRename ||
+                uniqueData->uniqueNameKind == core::UniqueNameKind::MangleRenameOverload ||
+                uniqueData->uniqueNameKind == core::UniqueNameKind::Overload) {
+                memberNameToHash = uniqueData->original;
+            }
+        }
 
-        // Because a change to classes would have take the slow path, should be safe
-        // to look up old owner in current foundDefs.
-        auto ownerSymbol = getOwnerSymbol(ownerRef);
-        ENFORCE(ownerSymbol.isClassOrModule());
-        auto owner = methodOwner(ctx, ownerSymbol, oldDefHash.owner.useSingletonClass);
+        auto fieldFullNameHash = core::FullNameHash(ctx, memberNameToHash);
+        return fieldFullNameHash == oldNameHash;
+    }
 
-        // We have to accumulate a list of methods to delete, instead of deleting them in the loop
-        // below, because deleting a method invalidates the members() iterator.
-        vector<core::MethodRef> toDelete;
+    void deleteSymbolViaFullNameHash(core::MutableContext ctx, core::ClassOrModuleRef owner,
+                                     core::FullNameHash oldNameHash) {
+        // We have to accumulate a list of fields to delete, instead of deleting them in the loop
+        // below, because deleting a field invalidates the members() iterator.
+        vector<core::SymbolRef> toDelete;
 
-        // Note: this loop is accidentally quadratic. We run deleteMethodViaFullNameHash once per method
-        // previously defined in this file, then in each call look at each member of that method's owner.
+        // Note: this loop is accidentally quadratic. We run deleteFieldViaFullNameHash once per field
+        // previously defined in this file, then in each call look at each member of that field's owner.
         for (const auto &[memberName, memberSym] : owner.data(ctx)->members()) {
-            if (!memberSym.isMethod()) {
+            if (memberSym.isClassOrModule()) {
+                // Safeguard against a collision function in our `FullNameHash` function (e.g., what
+                // if `foo()` and `::Foo` have the same `FullNameHash`, but were given two different
+                // `NameRef` IDs and thus don't collide in the `members` list?).
+                //
+                // Any other collisions are fine, because we will already be re-entering the new
+                // definitions later on in incremental namer. It's only classes that we definitely
+                // never want to delete accidentally.
```

Comment:
I was going to write something up about how this check doesn't seem sufficient, but I guess the argument is that we were going to change the locs anyway, this is just doing it a bit earlier than we were before.